service/agentconf.go

```diff
@@ -192,23 +192,34 @@ func (s *systemdServiceManager) WriteSystemdAgents(
 		if !ok {
 			return nil, nil, nil, errors.Errorf("%s service not of type UpgradableService", svcName)
 		}
+
+		dbusMethodNotFound := false
 		if err = uSvc.WriteService(); err != nil {
-			logger.Infof("failed to write service for %s: %s", agentName, err)
-			errAgentNames = append(errAgentNames, agentName)
-			lastError = err
-			continue
+			// Note that this error is already logged by the systemd package.
+
+			// This is not ideal, but it is possible on an Upstart-based OS
+			// (such as Trusty) for run/systemd/system to exist, which is used
+			// for detection of systemd as the running init system.
+			// If this happens, then D-Bus will error with the message below.
+			// We need to detect this condition and fall through to linking the
+			// service files manually.
+			if strings.Contains(err.Error(), "No such method 'LinkUnitFiles'") {
+				dbusMethodNotFound = true
+				logger.Infof("attempting to manually link service file for %s", agentName)
+			} else {
+				errAgentNames = append(errAgentNames, agentName)
+				lastError = err
+				continue
+			}
 		} else {
 			logger.Infof("successfully wrote service for %s:", agentName)
 		}
 
-		running, err := s.isRunning()
-		switch {
-		case err != nil:
-			return nil, nil, nil, errors.Errorf("failed to determine if systemd is running: %#v\n", err)
-		case running:
-			// If systemd is the running init system on this host, then the
-			// call to DBusAPI.LinkUnitFiles in WriteService above will have
-			// automatically updated sym-links for the file. We are done.
+		// If systemd is the running init system on this host, *and* if the
+		// call to DBusAPI.LinkUnitFiles in WriteService above returned no
+		// error, it will have resulted in updated sym-links for the file.
+		// We are done.
+		if s.isRunning() && !dbusMethodNotFound {
```

Comment:
Using double negatives is a bit hard to follow here. Can we just use:
dbusMethodMissing
or 
dbusMethodFound  and invert the logic?

Reply:
Done.